src/lazy/expanded/macro_evaluator.rs

```diff
@@ -273,22 +408,34 @@ impl<
             };
 
             // Ask that expansion to continue its evaluation by one step.
-            use MacroExpansionStep::*;
-            match current_expansion.next(context)? {
+            use ValueExpr::*;
+            match current_expansion.next(context, environment)? {
                 // If we get a value, return it to the caller.
-                ExpandedValue(value) => return Ok(Some(value)),
+                Some(ValueLiteral(value)) => {
+                    return Ok(Some(value));
+                }
                 // If we get another macro, push it onto the stack and continue evaluation.
-                AnotherMacroToEvaluate(invocation) => {
+                Some(MacroInvocation(invocation)) => {
                     // If we encounter another macro invocation, put it on top of the stack.
                     self.push(context, invocation)?;
                     continue;
                 }
                 // If the current macro reports that its expansion is complete...
-                Complete => {
-                    // ...pop it off the stack...
-                    let _popped = self.macro_stack.pop().unwrap();
+                None => {
+                    // Check to see if the completed value was a template. If so, discard its environment.
+                    let completed_kind = &self.macro_stack.last().unwrap().kind;
+                    if matches!(completed_kind, MacroExpansionKind::Template(_)) {
+                        // NB: Here and below, we use `truncate()` instead of `pop()` so the value can
+                        // be dropped in place without incurring a move. That move runs afoul of the
+                        // aliasing requirements that `miri` looks for, though I'm unsure why.
+                        // Once Polonius lands and we are able to remove the `unsafe` usages in
+                        // the LazyExpandingReader, this will be unnecessary.
+                        self.env_stack.truncate(self.env_stack.len() - 1);
+                    }
+                    self.macro_stack.truncate(self.macro_stack.len() - 1);
```

Comment:
🗺️ There is some `unsafe` code in the `LazyExpandingReader` that exists purely as a workaround to a limitation in `rustc` that will be fixed some time in the coming months. `miri` reported that `pop()` was problematic, but I'm skeptical that it mattered. Running `miri` again with its more sophisticated [tree-borrow model](https://perso.crans.org/vanille/treebor/) enabled allowed it to pass without complaint.